Generate P/Invoke transitions for CoreRT.

pgavlin opened the pull request:
This change adds support for CoreRT-style P/Invoke transitions
to RyuJIT. Instead of the usual inlined transition code, these
transitions are made up of calls to two new JIT helpers:

    PInvokeTransitionFrame frame; // opaque local
    CORINFO_HELP_INIT_PINVOKE_FRAME(&frame);
    ...
    CORINFO_HELP_JIT_PINVOKE_BEGIN(&frame);
    target(...);
    CORINFO_HELP_JIT_PINVOKE_END(&frame);
    ...

The preemptive mode constraints apply between calls to
JIT_PINVOKE_BEGIN and JIT_PINVOKE_END: no managed references
may be live only in registers and managed references may not
be manipulated.

pgavlin commented:
@jkotas @dotnet/jit-contrib PTAL. There are some corresponding changes that need to be made in CoreRT that I'll post later this weekend.

pgavlin commented on src/inc/corjit.h:
```diff
@@ -124,17 +124,18 @@ enum CorJitFlag
+    CORJIT_FLG_PINVOKE_USE_HELPERS = 0x00004000, // Use JIT_PINVOKE_{BEGIN,END} helpers instead of generating transitions inline.
+
+    // TODO: this should probably not overlap with CORJIT_FLG_MAKEFINALCODE, but we've run out of bits.
+    CORJIT_FLG_PINVOKE_DIRECT_CALLS= 0x00008000, // Direct calls to native methods do not require indirections
```
FWIW, I would like to figure out some alternative to this second flag before checking in. I had originally planned on using `CORJIT_FLG_UNUSED1`, but that bit is not available because RyuJit builds with `FEATURE_LEGACYNETCF` enabled. Any suggestions are welcome.

jkotas replied:
You should just add a new flag to CorJitFlag2.

pgavlin replied:
AFAICT the values in `CorJitFlag2` are never passed to the JIT. Only the VM seems to have access to them. I think your later suggestion to change the signature of `getAddressOfPInvokeFixup` is better in any case.

CarolEidt commented on src/jit/lclvars.cpp:
```diff
@@ -2159,17 +2159,24 @@ void               Compiler::lvaDecRefCnts(GenTreePtr tree)
     {
-        /* Get the special variable descriptor */
+        if (opts.eeFlags & CORJIT_FLG_PINVOKE_USE_HELPERS)
+        {
```
It doesn't really seem necessary to have noway_asserts executed this frequently. I would think that leaving only one such check prior to Lowering, say, would be sufficient.

pgavlin replied:
That makes sense. Would you object to my changing these to normal asserts?

CarolEidt commented:
One minor comment; otherwise LGTM.

sivarv commented on src/jit/codegencommon.cpp:
```diff
@@ -7557,6 +7557,7 @@ void CodeGen::genPrologPadForReJit()
 regMaskTP           CodeGen::genPInvokeMethodProlog(regMaskTP initRegs)
 {
     assert(compiler->compGeneratingProlog);
+    noway_assert((compiler->opts.eeFlags & CORJIT_FLG_PINVOKE_USE_HELPERS) == 0);
```
Rather than duplicating this check at multiple places, i would suggest to abstract it in a routine in Compiler::Options struct as  shouldUsePInvokeHelpers()

e.g. if (compiler->opts.shouldUsePInvokeHelpers())

sivarv replied:
shouldUsePInvokeHelpers() could be defined to return true only for CoreRT and false in all other cases.

pgavlin replied:
Yes, I thought about this, but it didn't seem worth it at the time. I'll go ahead and make the suggested change.

sivarv commented on src/inc/corinfo.h:
```diff
@@ -695,6 +695,9 @@ enum CorInfoHelpFunc
     CORINFO_HELP_THROW_ARGUMENTOUTOFRANGEEXCEPTION, // throw ArgumentOutOfRangeException
 #endif
 
+    CORINFO_HELP_JIT_PINVOKE_BEGIN, // Transition to preemptive mode before a P/Invoke, frame is the first argument
```
I believe these are used only by Core RT right?  In such a case should these be constrained to be available when compiling Core RT?

pgavlin replied:
See Jan's comment below.

sivarv commented on src/inc/corjit.h:
```diff
@@ -124,17 +124,18 @@ enum CorJitFlag
     CORJIT_FLG_NO_MDIL             = 0x00010000, // Generate an MDIL section but no code or CTL. Not used by the JIT, used internally by NGen only.
-#endif // MDIL
+#else // MDIL
+    CORJIT_FLG_PINVOKE_USE_HELPERS = 0x00004000, // Use JIT_PINVOKE_{BEGIN,END} helpers instead of generating transitions inline.
```
If we constrain this flag to enabled only when compiling Core RT, then we can reuse LegacyNetCF flag because Core RT doesn't have to care about NetCF compat.

pgavlin replied:
See Jan's comment below.

sivarv commented on src/inc/jithelpers.h:
```diff
@@ -403,6 +403,9 @@
 
 #endif // COR_JIT_EE_VERSION
 
+    JITHELPER1(CORINFO_HELP_JIT_PINVOKE_BEGIN, NULL, CORINFO_HELP_SIG_UNDEF, MDIL_HELP_UNDEF)
```
Should we constrain these under a #ifdef  so that they get included when compiling Core RT?

jkotas replied:
The JIT-EE interface should be consistent accross runtimes. It means that each runtime may pay a tiny price for it by having unused bits here and there, but this price is worth paying to avoid ifdef hell.

sivarv commented on src/jit/codegencommon.cpp:
```diff
@@ -8517,13 +8519,23 @@ void                CodeGen::genFnProlog()
+        excludeMask |= RBM_PINVOKE_FRAME;
+        if (compiler->opts.eeFlags & CORJIT_FLG_PINVOKE_USE_HELPERS)
+        {
+            noway_assert(compiler->info.compLvFrameListRoot == BAD_VAR_NUM);
```
It would be a good idea to update the comment in compiler.h where compLvFrameListRoot field is declared to indicate that this is not used by Core RT.

pgavlin replied:
Sure.

sivarv commented on src/jit/codegencommon.cpp:
```diff
@@ -8517,13 +8519,23 @@ void                CodeGen::genFnProlog()
+        }
+        else
+        {
+            noway_assert(compiler->info.compLvFrameListRoot < compiler->lvaCount);
```
IMO, a noway_assert looks too strong.  Here an assert will do.

sivarv replied:
A noway assert will be present not just in debug even in retail code.  Further noway assert when fired in retail will trigger JIT compilation in minopts.  If the same noway assert is hit even in min opts then it will lead to throwing an Invalid program exception.

pgavlin replied:
Aha, I was wondering what exactly the difference was. I'll admit to pattern-matching a bit here against other `noway_assert`s that ensure that there is a frame list root variable.

sivarv commented on src/jit/flowgraph.cpp:
```diff
@@ -8249,7 +8253,7 @@ void                Compiler::fgAddInternal()
         // Grab and reserve space for TCB, Frame regs used in PInvoke epilog to pop the inlined frame.
         // See genPInvokeMethodEpilog() for use of the grabbed var.
-        if (compJmpOpUsed)
+        if (!usePInvokeHelpers && compJmpOpUsed)
```
It would be a good idea to add a comment why we are excluding lvaPInvokeFrameRegSaveVar  for Core RT.

sivarv commented on src/jit/lclvars.cpp:
```diff
@@ -2159,17 +2159,24 @@ void               Compiler::lvaDecRefCnts(GenTreePtr tree)
 
-        varDsc->decRefCnts(compCurBB->getBBWeight(this), this);  
-        varDsc->decRefCnts(compCurBB->getBBWeight(this), this);
+            noway_assert(lclNum <= lvaCount);
```
IMO, this noway assert is too strong for a frequently executed code in coreclr JIT

pgavlin replied:
I did not add this assert.

briansull replied:
This assert only executes once for each Unmanaged Calls, which is not frequently.

jkotas commented on src/jit/lower.cpp:
```diff
@@ -2809,7 +2860,7 @@ GenTree* Lowering::LowerNonvirtPinvokeCall(GenTreeCall* call)
     InsertPInvokeCallProlog(call);
 
-    if (call->gtCallType != CT_INDIRECT)
+    if ((comp->opts.eeFlags & CORJIT_FLG_PINVOKE_DIRECT_CALLS) == 0 && call->gtCallType != CT_INDIRECT)
```
Instead of adding the `CORJIT_FLG_PINVOKE_DIRECT_CALLS` flag, it would be better (consistent with the JIT/EE design) to change `getAddressOfPInvokeFixup` to return CORINFO_CONST_LOOKUP. How many indirections are needed can be then described via `IAT_VALUE/IAT_PVALUE/IAT_PPVALUE`.

(If there is a problem with changing getAddressOfPInvokeFixup, add a new method to the JIT-EE interface that does the right thing.)

sivarv replied:
Why checking for CORJIT_FLG_PINVOKE_DIRECT_CALLS flag?
Btw, in my source search i couldn't find this flag anywhere.  What am i missing?

pgavlin replied:
@jkotas good suggestion. I didn't know that I had to bump the JIT/EE interface version for this change since I was only adding values to enums so I did not even consider this. If I have to bump the version anyway, I think this is the right direction to go.

@sivarv: the flag was added with this change. This flag is checked because there is no need for an indirection when performing direct calls to native methods in CoreRT.

jkotas commented on src/inc/corinfo.h:
```diff
@@ -695,6 +695,9 @@ enum CorInfoHelpFunc
     CORINFO_HELP_THROW_ARGUMENTOUTOFRANGEEXCEPTION, // throw ArgumentOutOfRangeException
 #endif
```
Bump the JIT-EE interface version at the top of the file?

pgavlin replied:
Okay. I wasn't sure if this was necessary since I was only adding enum values.

jkotas commented:
The JIT-EE interface will need matching fixes in the internal build to avoid build breaks. I think it would be better to do it in TFS.

jkotas commented on src/vm/jitinterface.cpp:
```diff
@@ -9953,8 +9953,8 @@ void* CEEInfo::getPInvokeUnmanagedTarget(CORINFO_METHOD_HANDLE method,
 /*********************************************************************/
     // return address of fixup area for late-bound N/Direct calls.
-void* CEEInfo::getAddressOfPInvokeFixup(CORINFO_METHOD_HANDLE method,
-                                        void **ppIndirection)
```
I think a better name for this method would be `getAddressOfPInvokeTarget` since the result may or may not be a fixup now.

Also, you can rename `getPInvokeUnmanagedTarget` to something like `getPInvokeUnmanagedTarget_Obsolete` while you are on it. In the ideal case, it would be deleted but I am not sure whether it is possible because of full .NET framework compat.

pgavlin replied:
> I think a better name for this method would be getAddressOfPInvokeTarget since the result may or may not be a fixup now.

@CarolEidt @BruceForstall @jkotas: I wonder if it would be worth combining `getAddressOfPInvokeFixup` and `getFunctionEntryPoint`. This may also allow us to share code between `Lowering::LowerNonvirtPinvokeCall` and `Lowering::LowerDirectCall`, which really only differ in that the former inserts a few P/Invoke-specific constructs around the call. If we were to do this, both `getAddressOfPInvokeFixup` and `getPInvokeUnmanagedTarget` would be dead code in CoreCLR and could be marked obsolete.

jkotas replied:
> getAddressOfPInvokeTarget  and  getFunctionEntryPoint 

These two methods return different entrypoints. I do not think they should be merged. (If they were merged, the merged method would need a flag which one of the two different entrypoints is needed ... .)

pgavlin replied:
> These two methods return different entrypoints

Can you break down the different entrypoints that would be returned by the two methods for a P/Invoke? In other words, given some P/Invoke P, what is the difference between a direct call to the result of `getFunctionEntryPoint(P)` and the result of `getAddressOfPInvokeFixup(P)`?

jkotas replied:
getFunctionEntryPoint returns the managed entrypoint. It is basically the result of ldftn `<method>`. It is valid for any method. For P/Invoke methods, it is address of the PInvoke marshalling stub.

 getAddressOfPInvokeFixup returns the unmanaged PInvoke target. It is the address of the actual unmanaged method that the PInvoke calls (e.g. `kernel32!ReadFile`).

pgavlin replied:
So does a direct P/Invoke always bypass marshalling?

jkotas replied:
Yes, the direct P/Invoke is only allowed when the marshalling is not required (e.g. when `pInvokeMarshalingRequired` returns false).

pgavlin replied:
Okay, I see now. I agree that the methods cannot be merged, then, and that we should rename the method as you suggest.

LLITCHEV commented on src/jit/lclvars.cpp:
```diff
@@ -2214,17 +2218,21 @@ void               Compiler::lvaIncRefCnts(GenTreePtr tree)
 
-        varDsc->incRefCnts(compCurBB->getBBWeight(this), this);  
-        varDsc->incRefCnts(compCurBB->getBBWeight(this), this);
+            noway_assert(lclNum <= lvaCount);
```
I think VarNums (lclNum is such) are 0 based. Should the noway_assert be  lclNum < lvaCount?

pgavlin replied:
This assertion was pre-existing; I'd prefer to leave it as is.

LLITCHEV replied:
Sure. Please add an issue, so it gets fixed at some point.

jkotas commented:
@kyulee1 Have you merged this by accident? My last round of feedback was not addressed yet, and it will need special handholding for internal build - as I have said above, the JIT-EE interface update should be better done from the TFS side.

kyulee1 commented:
@jkotas I don't even remember I clicked the merge. Maybe I did it for another PR.

pgavlin commented:
@jkotas: how do you want to move forward here? Should I revert this change and continue addressing feedback here, or should I submit a new PR or change in TFS to address feedback?
